src/usexp/atom.ml

```diff
@@ -0,0 +1,49 @@
+type t = A of string [@@unboxed]
+
+let invalid_argf fmt = Printf.ksprintf invalid_arg fmt
+
+type syntax = Jbuild | Dune
+
+let (is_valid_jbuild, is_valid_dune) =
+  let rec jbuild s i len =
+    i = len ||
+    match String.unsafe_get s i with
+    | '#' -> disallow_next '|' s (i + 1) len
+    | '|' -> disallow_next '#' s (i + 1) len
+    | '"' | '(' | ')' | ';' | '\000'..'\032' | '\127'..'\255' -> false
```

Comment:
Some of these characters where allowed in jbuild syntax. Let's just restore the old function:

```ocaml
let is_valid str =
  let len = String.length str in
  len > 0 &&
  let rec loop ix =
    match str.[ix] with
    | '"' | '(' | ')' | ';' -> true
    | '|' -> ix > 0 && let next = ix - 1 in str.[next] = '#' || loop next
    | '#' -> ix > 0 && let next = ix - 1 in str.[next] = '|' || loop next
    | ' ' | '\t' | '\n' | '\012' | '\r' -> true
    | _ -> ix > 0 && loop (ix - 1)
  in
  not (loop (len - 1))
```

Reply:
The old function was this:

```
 let is_valid =
   let rec loop s i len =
     i = len ||
     match String.unsafe_get s i with
     | '"' | '(' | ')' | ';' | '\000'..'\032' | '\127'..'\255' -> false
     | _ -> loop s (i + 1) len
   in
   fun s ->
     let len = String.length s in
     len > 0 && loop s 0 len
```

I'll use your definition however.

Reply:
I meant from before #837